File: app/src/main/java/com/scullyapps/spoonsbottleup/SettingsActivity.java
package com.scullyapps.spoonsbottleup;

import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;

import com.google.android.material.tabs.TabLayout;

import androidx.appcompat.widget.Toolbar;
import androidx.viewpager.widget.ViewPager;
import androidx.appcompat.app.AppCompatActivity;

import android.view.MenuItem;
import android.widget.Toast;

import com.scullyapps.spoonsbottleup.ui.Fridge;
import com.scullyapps.spoonsbottleup.ui.settings.BottleListFragment;
import com.scullyapps.spoonsbottleup.ui.settings.FridgeFragment;
import com.scullyapps.spoonsbottleup.ui.settings.GeneralSettingsFragment;
import com.scullyapps.spoonsbottleup.ui.settings.SectionsPagerAdapter;

public class SettingsActivity extends AppCompatActivity implements BottleListFragment.OnFragmentInteractionListener, FridgeFragment.OnListFragmentInteractionListener {

    private int CURRENT_TAB = 0;

    private Toolbar toolbar;
    private Context context;
    private TabLayout  tabs;
    private ViewPager viewPager;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        setContentView(R.layout.activity_settings);
        SectionsPagerAdapter sectionsPagerAdapter = new SectionsPagerAdapter(this, getSupportFragmentManager());

        viewPager = findViewById(R.id.view_pager);
        viewPager.setAdapter(sectionsPagerAdapter);

        tabs = findViewById(R.id.tabs);
        tabs.setupWithViewPager(viewPager);

        toolbar = findViewById(R.id.toolbar_settings);
        context = this;

        setupListeners();

        getSupportFragmentManager()
                .beginTransaction()
                .replace(R.id.view_pager, new GeneralSettingsFragment())
                .commit();

    }

    @Override
    public void onFragmentInteraction(Uri uri) {

    }


    @Override
    public void onListFragmentInteraction(Bottle item) {
        Toast.makeText(this, "Item Pressed " + item.getName(), Toast.LENGTH_SHORT).show();
    }

    @Override
    public void onListFragmentInteraction(Fridge item) {
        Intent i = new Intent(this, FridgeManagementActivity.class);
        i.putExtra("name", item.getName());

        startActivity(i);

    }

    private void setupListeners() {
        // menu options
        toolbar.setOnMenuItemClickListener(new Toolbar.OnMenuItemClickListener() {
            @Override
            public boolean onMenuItemClick(MenuItem item) {

                switch (item.getItemId()) {
                    case R.id.action_add_bottle:
                        Toast.makeText(context, "Pressed adding fridge", Toast.LENGTH_LONG).show();
                        break;

                    case R.id.action_add_fridge:
                        Toast.makeText(context, "Pressed adding fridge", Toast.LENGTH_LONG).show();
                        break;
                }

                return false;
            }
        });

        tabs.addOnTabSelectedListener(new TabLayout.OnTabSelectedListener() {
            @Override
            public void onTabSelected(TabLayout.Tab tab) {

                // update class-wide tab selection
                CURRENT_TAB = tab.getPosition();

                // 0 - general
                // 1 - bottles
                // 2 - fridges

                switch (CURRENT_TAB) {
                    case 0:

                        break;

                    case 1:
                        toolbar.inflateMenu(R.menu.menu_settings_bottles);
                        break;

                    case 2:
                        toolbar.inflateMenu(R.menu.menu_settings_fridges);
                        break;
                }
            }

            @Override
            public void onTabUnselected(TabLayout.Tab tab) {
                // remove previously inflated menus
                toolbar.getMenu().clear();
            }

            @Override public void onTabReselected(TabLayout.Tab tab) { }
        });
    }
}

File: app/src/main/java/com/scullyapps/spoonsbottleup/activities/SettingsActivity.kt
package com.scullyapps.spoonsbottleup.activities

import android.app.AlertDialog
import android.content.Intent
import android.os.Bundle
import android.text.InputType
import android.widget.EditText
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import com.google.android.material.tabs.TabLayout
import com.google.android.material.tabs.TabLayout.OnTabSelectedListener
import com.scullyapps.spoonsbottleup.R
import com.scullyapps.spoonsbottleup.adapters.SettingsPagerAdapter
import com.scullyapps.spoonsbottleup.data.BottleDatabase
import com.scullyapps.spoonsbottleup.fragments.FridgeFragment
import com.scullyapps.spoonsbottleup.fragments.FridgeFragment.OnListFragmentInteractionListener
import com.scullyapps.spoonsbottleup.fragments.GeneralSettingsFragment
import com.scullyapps.spoonsbottleup.models.Bottle
import com.scullyapps.spoonsbottleup.models.Fridge
import com.scullyapps.spoonsbottleup.ui.FridgeView
import kotlinx.android.synthetic.main.activity_settings.*

class SettingsActivity : AppCompatActivity(), OnListFragmentInteractionListener {
    private var CURRENT_TAB = 0

    lateinit var toolbar: Toolbar
    lateinit var sectionsPagerAdapter : SettingsPagerAdapter

    lateinit var database: BottleDatabase

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_settings)

        database = BottleDatabase.getInstance(this)

        sectionsPagerAdapter = SettingsPagerAdapter(this, supportFragmentManager)
        toolbar = findViewById(R.id.toolbar_settings)

        view_pager.adapter = sectionsPagerAdapter

        tabs.setupWithViewPager(view_pager)

        setupListeners()

        supportFragmentManager
                .beginTransaction()
                .replace(R.id.view_pager, GeneralSettingsFragment())
                .commit()
    }

    override fun onListFragmentInteraction(item: Bottle?) {
        Toast.makeText(this, "Item Pressed " + item?.name, Toast.LENGTH_SHORT).show()
    }

    override fun onListFragmentInteraction(item: FridgeView?, action: FridgeFragment.InteractionAction) {

        // do nothing if item null
        if(item == null)
            return

        when (action) {
            // Delete from database, update recycler
            FridgeFragment.InteractionAction.DELETE -> {

                val builder = AlertDialog.Builder(this).apply {
                    setTitle("Delete fridge")
                    setMessage("Are you sure you wish to delete fridge: " + item.fridge.name + "?")
                    setPositiveButton("Delete") {_, _ ->
                        database.fridgeRoomDao.delete(item.fridge)
                        getFridgeFragment()?.refreshRecyclerView(this@SettingsActivity)
                    }
                }.create().show()
            }

            // Open fridge edit activity
            FridgeFragment.InteractionAction.EDIT -> {
                val intent = Intent(this, FridgeManagementActivity::class.java).apply {
                    putExtra("name", item.fridge.name)
                }

                startActivity(intent)
            }
        }
    }

    private fun getFridgeFragment() : FridgeFragment? {
        val key = SettingsPagerAdapter.KEY_FRIDGE_FRAGMENT

        return if(sectionsPagerAdapter.hasFragment(key)) {
            sectionsPagerAdapter.getFragment(key) as FridgeFragment
        } else {
            null
        }
    }

    private fun addFridge() {
        AlertDialog.Builder(this).apply {
            val editText = EditText(this@SettingsActivity).apply {
                inputType = InputType.TYPE_CLASS_TEXT
            }

            setView(editText)
            setTitle("Add a new fridge")

            setPositiveButton("Add") { _, _ ->
                database.fridgeRoomDao.insert(Fridge(0, editText.text.toString(), 0))
                getFridgeFragment()?.refreshRecyclerView(this@SettingsActivity)
            }

        }.create().show()
    }

    private fun setupListeners() {
        // menu options
        toolbar.setOnMenuItemClickListener { item ->
            when (item.itemId) {
                R.id.action_add_fridge -> addFridge()
            }
            false
        }

        tabs!!.addOnTabSelectedListener(object : OnTabSelectedListener {
            override fun onTabSelected(tab: TabLayout.Tab) {
                // update class-wide tab selection
                CURRENT_TAB = tab.position
                when (CURRENT_TAB) {
                    0 -> {}
                    1 -> toolbar.inflateMenu(R.menu.menu_settings_fridges)
                }
            }

            override fun onTabUnselected(tab: TabLayout.Tab) {
                // remove previously inflated menus
                toolbar.menu.clear()
            }

            override fun onTabReselected(tab: TabLayout.Tab) {}
        })
    }



}